lib/pbench/server/auth/auth.py

```diff
@@ -130,36 +190,58 @@ def verify_auth(auth_token):
         return None
 
     @staticmethod
-    def verify_third_party_token(auth_token: str, oidc_client: OpenIDClient) -> bool:
+    def verify_third_party_token(
+        auth_token: str,
+        oidc_client: OpenIDClient,
+        algorithms: Optional[list[str]] = ["HS256"],
+    ) -> "InternalUser":
         """
         Verify a token provided to the Pbench server which was obtained from a
         third party identity provider.
         Args:
             auth_token: Token to authenticate
             oidc_client: OIDC client to call to authenticate the token
+            algorithms: Optional token signature algorithm argument,
+                        defaults to HS256
         Returns:
             True if the verification succeeds else False
         """
-        identity_provider_pubkey = oidc_client.get_oidc_public_key(auth_token)
         try:
-            oidc_client.token_introspect_offline(
+            identity_provider_pubkey = oidc_client.get_oidc_public_key(auth_token)
+        except Exception:
+            Auth.logger.warning("Identity provider public key fetch failed")
+            identity_provider_pubkey = Auth().get_secret_key()
+        try:
+            audience = oidc_client.client_id if oidc_client else None
+            token_decode = oidc_client.token_introspect_offline(
                 token=auth_token,
                 key=identity_provider_pubkey,
-                audience=oidc_client.client_id,
+                audience=audience,
+                algorithms=algorithms,
                 options={
                     "verify_signature": True,
                     "verify_aud": True,
                     "verify_exp": True,
                 },
             )
-            return True
+            roles = []
+            if audience in token_decode.get("resource_access"):
+                roles = token_decode.get("resource_access").get(audience).get("roles")
+            return InternalUser(
+                id=token_decode.get("sub"),
+                username=token_decode.get("preferred_username"),
+                email=token_decode.get("email"),
+                first_name=token_decode.get("given_name"),
+                last_name=token_decode.get("family_name"),
+                roles=roles,
+            )
         except (
             jwt.ExpiredSignatureError,
             jwt.InvalidTokenError,
             jwt.InvalidAudienceError,
         ):
-            Auth.logger.error("OIDC token verification failed")
-            return False
+            Auth.logger.warning("OIDC token verification failed")
```

Comment:
I don't think we want to log authorization failures. The API will return a 401, and that's all we care about. (We'll have the Apache/gunicorn logs to track failed HTTP requests if we really need that.)

Reply:
Yes, I added this when I was testing this code.